// @sara-via/server/src/auth/auth.service.ts
import crypto from 'crypto';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { PiAuthStrategy } from './pi-auth.strategy';
import { PrismaService } from '../prisma/prisma.service';

interface RequestContextMeta {
  ipAddress?: string;
  userAgent?: string;
}

@Injectable()
export class AuthService {
  constructor(
    private readonly piAuthStrategy: PiAuthStrategy,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
    private readonly prisma: PrismaService
  ) {}

  private hashToken(token: string): string {
    return crypto.createHash('sha256').update(token).digest('hex');
  }

  private async issueTokenPair(
    user: { uid: string; username?: string },
    meta: RequestContextMeta,
    existingSessionId?: string
  ) {
    const jwtConfig = this.configService.getOrThrow('app.jwt');
    const refreshTokenTtlDays = this.configService.getOrThrow<number>('app.refreshTokenTtlDays');
    const sessionId = existingSessionId ?? crypto.randomUUID();
    const basePayload = {
      sub: user.uid,
      uid: user.uid,
      username: user.username,
      sid: sessionId
    };

    const [accessToken, refreshToken] = await Promise.all([
      this.jwtService.signAsync(basePayload, {
        algorithm: 'RS256',
        privateKey: jwtConfig.privateKey,
        issuer: jwtConfig.issuer,
        audience: jwtConfig.audience,
        expiresIn: '15m'
      }),
      this.jwtService.signAsync(
        { ...basePayload, type: 'refresh' },
        {
          algorithm: 'RS256',
          privateKey: jwtConfig.privateKey,
          issuer: jwtConfig.issuer,
          audience: jwtConfig.audience,
          expiresIn: `${refreshTokenTtlDays}d`
        }
      )
    ]);

    const refreshTokenHash = this.hashToken(refreshToken);
    const expiresAt = new Date(Date.now() + refreshTokenTtlDays * 24 * 60 * 60 * 1000);

    await this.prisma.refreshTokenSession.upsert({
      where: { id: sessionId },
      create: {
        id: sessionId,
        pioneerUid: user.uid,
        refreshTokenHash,
        userAgent: meta.userAgent,
        ipAddress: meta.ipAddress,
        expiresAt
      },
      update: {
        refreshTokenHash,
        userAgent: meta.userAgent,
        ipAddress: meta.ipAddress,
        expiresAt,
        revokedAt: null
      }
    });

    return { accessToken, refreshToken, user };
  }

  async loginWithPi(accessToken: string, meta: RequestContextMeta = {}) {
    const user = await this.piAuthStrategy.validateAccessToken(accessToken);
    return this.issueTokenPair(user, meta);
  }

  async refreshSession(refreshToken: string, meta: RequestContextMeta = {}) {
    const jwtConfig = this.configService.getOrThrow('app.jwt');
    const payload = await this.jwtService.verifyAsync<{
      uid: string;
      username?: string;
      sid: string;
      type: string;
    }>(refreshToken, {
      algorithms: ['RS256'],
      publicKey: jwtConfig.publicKey,
      issuer: jwtConfig.issuer,
      audience: jwtConfig.audience
    });

    if (payload.type !== 'refresh') {
      throw new UnauthorizedException('Provided token is not a refresh token');
    }

    const session = await this.prisma.refreshTokenSession.findUnique({ where: { id: payload.sid } });
    if (!session || session.pioneerUid !== payload.uid || session.revokedAt || session.expiresAt.getTime() < Date.now()) {
      throw new UnauthorizedException('Refresh session is invalid or expired');
    }

    const incomingHash = this.hashToken(refreshToken);
    if (incomingHash !== session.refreshTokenHash) {
      await this.prisma.refreshTokenSession.update({
        where: { id: session.id },
        data: { revokedAt: new Date() }
      });
      throw new UnauthorizedException('Refresh token rotation detected a replay attempt');
    }

    return this.issueTokenPair(
      { uid: payload.uid, username: payload.username },
      meta,
      session.id
    );
  }
}
