// @sara-via/server/src/payments/pi-network.service.ts
import { BadGatewayException, Injectable, UnauthorizedException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { PiPaymentDto } from './schemas/transaction.schema';

@Injectable()
export class PiNetworkService {
  private readonly piApiBaseUrl: string;
  private readonly piApiKey: string;

  constructor(private readonly configService: ConfigService) {
    this.piApiBaseUrl = this.configService.getOrThrow<string>('app.piApiBaseUrl');
    this.piApiKey = this.configService.getOrThrow<string>('app.piApiKey');
  }

  private async request<T>(path: string, init: RequestInit): Promise<T> {
    const response = await fetch(`${this.piApiBaseUrl}${path}`, init);

    if (!response.ok) {
      const errorText = await response.text();
      throw new BadGatewayException(`Pi API ${path} failed: ${errorText}`);
    }

    return response.json() as Promise<T>;
  }

  async getCurrentUser(accessToken: string): Promise<{ uid: string; username?: string }> {
    const response = await fetch(`${this.piApiBaseUrl}/me`, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${accessToken}`
      }
    });

    if (response.status === 401) {
      throw new UnauthorizedException('Pi access token is invalid');
    }

    if (!response.ok) {
      throw new BadGatewayException(`Pi /me failed with status ${response.status}`);
    }

    return response.json() as Promise<{ uid: string; username?: string }>;
  }

  async getPayment(paymentId: string): Promise<PiPaymentDto> {
    return this.request<PiPaymentDto>(`/payments/${paymentId}`, {
      method: 'GET',
      headers: {
        Authorization: `Key ${this.piApiKey}`
      }
    });
  }

  async approvePayment(paymentId: string): Promise<PiPaymentDto> {
    return this.request<PiPaymentDto>(`/payments/${paymentId}/approve`, {
      method: 'POST',
      headers: {
        Authorization: `Key ${this.piApiKey}`
      }
    });
  }

  async completePayment(paymentId: string, txid: string): Promise<PiPaymentDto> {
    return this.request<PiPaymentDto>(`/payments/${paymentId}/complete`, {
      method: 'POST',
      headers: {
        Authorization: `Key ${this.piApiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ txid })
    });
  }
}
